src/events.jl

```diff
@@ -0,0 +1,302 @@
+    """
+    dropped_files::Observable{Vector{String}}
+    """
+    Whether the Scene window is in focus or not.
+    """
+    hasfocus::Observable{Bool}
+    """
+    Whether the mouse is inside the window or not.
+    """
+    entered_window::Observable{Bool}
+end
+
+function Base.show(io::IO, events::Events)
+    println(io, "Events:")
+    fields = propertynames(events)
+    maxlen = maximum(length ∘ string, fields)
+    for field in propertynames(events)
+        pad = maxlen - length(string(field)) + 1
+        println(io, "  $field:", " "^pad, to_value(getproperty(events, field)))
+    end
+end
+
+function Events()
+    events = Events(
+        Observable(Recti(0, 0, 0, 0)),
+        Observable(100.0),
+        Observable(false),
+
+        Observable(MouseButtonEvent(Mouse.none, Mouse.release)),
+        Set{Mouse.Button}(),
+        Observable((0.0, 0.0)),
+        Observable((0.0, 0.0)),
+
+        Observable(KeyEvent(Keyboard.unknown, Keyboard.release)),
+        Set{Keyboard.Button}(),
+        Observable('\0'),
+        Observable(String[]),
+        Observable(false),
+        Observable(false),
+    )
+
+    connect_states!(events)
+    return events
+end
+
+function connect_states!(e::Events)
+    on(e.mousebutton, priority = typemax(Int)) do event
+        set = e.mousebuttonstate
+        if event.action == Mouse.press
+            push!(set, event.button)
+        elseif event.action == Mouse.release
+            delete!(set, event.button)
+        else
+            error("Unrecognized Keyboard action $(event.action)")
+        end
+        # This never consumes because it just keeps track of the state
+        return Consume(false)
+    end
+
+    on(e.keyboardbutton, priority = typemax(Int)) do event
+        set = e.keyboardstate
+        if event.key != Keyboard.unknown
+            if event.action == Keyboard.press
+                push!(set, event.key)
+            elseif event.action == Keyboard.release
+                delete!(set, event.key)
+            elseif event.action == Keyboard.repeat
+                # set should already have the key
+            else
+                error("Unrecognized Keyboard action $(event.action)")
+            end
+        end
+        # This never consumes because it just keeps track of the state
+        return Consume(false)
+    end
+    return
+end
+
+# # Compat only
+# function Base.getproperty(e::Events, field::Symbol)
+#     if field === :mousebuttons
+#         error("`events.mousebuttons` is deprecated. Use `events.mousebutton` to react to `MouseButtonEvent`s instead.")
+#     elseif field === :keyboardbuttons
+#         error("`events.keyboardbuttons` is deprecated. Use `events.keyboardbutton` to react to `KeyEvent`s instead.")
+#     elseif field === :mousedrag
+#         error("`events.mousedrag` is deprecated. Use `events.mousebutton` or a mouse state machine (`addmouseevents!`) instead.")
+#     else
+#         return getfield(e, field)
+#     end
+# end
```

Comment:
This was just there to give people a useful error message when updating from an older version of the `Events` struct. I think it's fine to delete this now?

Reply:
2 years since those fieldnames were changed, 8 months since the warning was turned into an error